pkg/controller/kubedirectorcluster/cluster.go

```diff
@@ -91,7 +101,17 @@ func (r *ReconcileKubeDirectorCluster) syncCluster(
 		}
 	}()
 
-	// We use a finalizer to ensure that only KubeDirector updates status
+	// For a new CR just update the status state/gen.
+	shouldProcessCR := r.handleNewCluster(reqLogger, cr)
+	if !shouldProcessCR {
+		return nil
+	}
```

Comment:
So, hmm. Why this new ordering? Let's chew on this.

One thing that seems like a problem here is that we now *won't* call EnsureClusterAppReference when the cluster is first seen (because shouldProcessCR will be false). 

Also: previously we had intentionally moved handleFinalizers up *before* handleNewCluster, because we needed to move deleteStatusGen into handleFinalizers. Here was the reasoning for that:

-----
...we now have the situation where we could get a poll-triggered handler for a KDC CR that has a deletion timestamp and no finalizer, which has no entry in our status gen cache. We need to make sure that is handled correctly, i.e. ignored

So in the KDC reconciler, we want to make sure we don't do any processing for a CR that has a deletion timestamp, beyond (if necessary) removing its finalizer and removing its status gen from the cache. It looks like handleFinalizers already takes care of that. However, the current ordering of the calls to handleNewCluster (used to be handleStatusGen) and handleFinalizers is unfortunate in that handleNewCluster would re-add the status gen back into the cache. I *believe* the simplest answer to that problem is to switch the order that those functions are called -- always do handleFinalizers first.
-----

Our churn in how we are invoking these functions may be a sign to revisit them a bit.

This are of code is also related to some changes that Swami made in PR #168. In Swami's PR, handleStatusGen on a new cluster needs to be done *before* we try to add a finalizer (for various reasons). So I was thinking that, when I merge in this operator-sdk branch, I would need to split up the handleFinalizers code into two pieces and do things in this order:
* first take care of the `cr.DeletionTimestamp != nil` part of handleFinalizers
* next, bail out of reconcile if `cr.DeletionTimestamp != nil` (is deleted cluster)
* now take care of handleNewCluster (status gen stuff)
* bail out of reconcile if handleNewCluster asks us to  (is new cluster)
* finally take care of the EnsureFinalizer part of handleFinalizers
* bail out of reconcile if EnsureFinalizer failed

Since those are my plans, I'm curious about why you saw a need to reorder handleFinalizers/handleNewCluster in this PR and whether that issue would affect those plans.

Finally, w.r.t. the cluster app reference stuff...

The app reference is mostly about making sure that the app can't go away after we've started working on the cluster. If it goes away in a little race window after admitting the cluster and before we start working on the cluster, fundamentally we only need to make sure that we fail gracefully. And I believe we would.

Going back to Swami's PR here though: with his change, the validator specifically records the namespace of the app CR in the status of the cluster CR. It would probably make sense to do EnsureClusterAppReference *there*, in the validator, when admitting the cluster CR. The only other place we need to do EnsureClusterAppReference would be in handleNewCluster when uptaking an existing cluster after KD restart.

I'm not 100% sure what this means for your PR vs. what I just need to do in the merge.

I can do a Skype call to chat further. On a customer call for a bit longer here, then free after.

Reply:
I took the following comment as the intended behavior, "For a new CR just update the status state/gen." I took that to mean the statusGen would be assigned and updated in the resource before anything else happens. The code was allowing the finalizer update before because of a side effect/bug with the new update code. When I fixed that issue the validator was rejecting the finalizer update. L I can play with order again to see if I can get the finalizers updated first. et's talk when you get a chance.

Reply:
OK. Gonna lunch real quick first.

Reply:
I've been playing with the order and I think I have it the way that you want it. One issue I ran into was that the default Status object was being assigned too early. Since the Status in the cr object was changing the Client().Update() call to add the finalizer was failing with the error that Update() can't update the Status. Please have a look at the new committed code and let me know if it's ok. Thanks ...